Source/WebCore/platform/audio/cocoa/AudioDecoderCocoa.cpp

```diff
@@ -0,0 +1,349 @@
+{
+    queueSingleton().dispatch([decoder = m_internalDecoder] {
+        decoder->close();
+    });
+}
+
+void InternalAudioDecoderCocoa::decompressedAudioOutputBufferCallback(void* object, CMBufferQueueTriggerToken)
+{
+    // We can only be called from the CoreMedia callback if we are still alive.
+    RefPtr decoder = static_cast<class InternalAudioDecoderCocoa*>(object);
+
+    queueSingleton().dispatch([weakDecoder = ThreadSafeWeakPtr { *decoder }] {
+        if (auto strongDecoder = weakDecoder.get())
+            strongDecoder->processedDecodedOutputs();
+    });
+}
+
+void InternalAudioDecoderCocoa::processedDecodedOutputs()
+{
+    if (m_isClosed)
+        return;
+    while (RetainPtr sample = converter()->takeOutputSampleBuffer()) {
+        Ref audioData = PlatformRawAudioData::create(MediaSampleAVFObjC::create(sample.get(), 0));
+        m_outputCallback(AudioDecoder::DecodedData { WTFMove(audioData) });
+    }
+}
+
+InternalAudioDecoderCocoa::InternalAudioDecoderCocoa(AudioDecoder::OutputCallback&& outputCallback)
+    : m_outputCallback(WTFMove(outputCallback))
+{
+}
+
+String InternalAudioDecoderCocoa::initialize(const String& codecName, const AudioDecoder::Config& config)
+{
+    auto result = AudioDecoderCocoa::isCodecSupported(codecName);
+    if (!result)
+        return result.error();
+
+    auto [codec, format] = *result;
+    if (!format) {
+        if (codec == kAudioFormatOpus && config.numberOfChannels >= 2)
+            return "Opus with more than 2 channels isn't supported"_s;
+
+        if (codec == kAudioFormatFLAC && config.description.empty())
+            return "Decoder config description for flac codec is mandatory"_s;
+
+        if (codec == 'vorb' && config.description.empty())
+            return "Decoder config description for vorbis codec is mandatory"_s;
+
+        mIsAAC = codec == kAudioFormatMPEG4AAC || codec == kAudioFormatMPEG4AAC_HE || codec == kAudioFormatMPEG4AAC_LD || codec == kAudioFormatMPEG4AAC_HE_V2 || codec == kAudioFormatMPEG4AAC_ELD;
+
+        AudioStreamBasicDescription absd { };
+        absd.mFormatID = codec;
+        // Attempt to create description with provided cookie.
+        bool succeeded = false;
+        if (config.description.size()) {
+            UInt32 size = sizeof(absd);
+            succeeded = PAL::AudioFormatGetProperty(kAudioFormatProperty_FormatInfo, config.description.size(), config.description.data(), &size, &absd) == noErr;
+        }
+        if (!succeeded) {
+            absd.mSampleRate = double(config.sampleRate);
+            absd.mChannelsPerFrame = uint32_t(config.numberOfChannels);
+        }
+        m_inputDescription = absd;
+    } else
+        m_inputDescription = CAAudioStreamDescription { double(config.sampleRate), uint32_t(config.numberOfChannels), *format, CAAudioStreamDescription::IsInterleaved::Yes };
+
+    m_inputFormatDescription = createAudioFormatDescription(*m_inputDescription, m_codecDescription.span());
```

Comment:
m_codecDescription is never used but here.
Should we remove it from now and introduce it when we need it.

Reply:
actually, it's a mistake to not use it, it should be first set to config.description